File: examples/src/_12_error_handling.py

def main():
    user_input: str = get_user_input();
    print("You typed: {}".format(user_input));


def get_user_input() -> str:
    string_from_stdin = ''
    print("Please type a number between 0-100 and press enter:")
    while True:
        string_from_stdin = input() 
        trimmed = string_from_stdin.strip()
        try: # try to convert to int
            parsed = int(trimmed)
            if parsed < 0 or parsed > 100:
                print("Please type a number between 0-100 and press enter:")
                continue
            else:
                return str(trimmed)
        except ValueError:
            print("Please type a number between 0-100 and press enter:");
            continue

if __name__ == "__main__":
    main()

class User():
    pass

import sqlite3

def print_user_name(user_id):
    user: User = get_user(user_id)
    if user: # Se o usuário existir...
        print(user.name)

def get_user(conn, user_id) -> User:
    c = conn.cursor()
    c.execute(f"SELECT * FROM users WHERE id = {user_id}")
    user: User = c.fetchone()
    return user

File: examples/src/_12_error_handling.rs
use std::io;

#[allow(dead_code)]
pub fn example() {
    let user_input: String = get_user_input();
    println!("You typed: {}", user_input);

    let string_to_parse = String::from("123");
    let parsed = parse_u32(string_to_parse);
    let unwrapped = parsed.unwrap();
    println!("\n\n\nParsed: {}", unwrapped);
}

fn get_user_input() -> String {
    let mut string_from_stdin = String::new();
    println!("Please type a number between 0-100 and press enter:");
    loop {
        // Ler do stdin e armazenar em string_from_stdin. O valor de retorno é um Result, mas
        // não estamos interessados no valor de retorno.
        io::stdin()
            .read_line(&mut string_from_stdin)
            .expect("Failed to read line");

        let trimmed = string_from_stdin.trim();
        let parsed = trimmed.parse::<u32>();
        match parsed {
            Ok(i) => {
                if i > 100 {
                    println!("Please type a number between 0-100 and press enter:");
                    string_from_stdin = String::new(); // reset the string (read_line appends to it)
                    continue;
                }
                return trimmed.to_string();
            }
            Err(..) => {
                println!("Please type a number between 0-100 and press enter:");
                string_from_stdin = String::new(); // reset the string (read_line appends to it)
                continue;
            }
        };
    }
}

#[allow(dead_code)]
fn parse_u32(input: String) -> Result<u32, ()> {
    match input.trim().parse::<u32>() {
        Ok(i) => Ok(i),
        Err(..) => Err(()),
    }
}
